### ISSUE TITLE
finetune experience or code

### ISSUE THREAD

**Issue by niatzt** [openai/CLIP#150]
I tried to finetune CLIP on my own small dataset. The results are not satisfying (probably overfitting). After checking the issues here, I notice other people also have similar problems. Can someone share his successful finetune experience or code.

**Comment by ABaldrati** [comment]
Hi @niatzt 
In my limited experience of tine-tuning with CLIP I have often noticed how this operation is extremely sensitive to various hyperparameters.
The advice I feel like giving you (which has worked in my case) is:

-  keep CLIP in evaluation mode even during training (i.e. keeping the normalization layers frozen)

- use very low learning rate (even reaching values like 1e-7 or 1e-8)

- If you want to fine-tune CLIP together with another network trained from scratch, use different learning rates (lower for CLIP and higher for the other network). You may also consider keeping CLIP frozen a few epochs.

I hope these tips can help you and that your fine-tuning goes successfully.
If you have more questions do not hesitate to ask

**Comment by Sierkinhane** [comment]
I have finetuned CLIP on PASCAL VOC2012 dataset using this [implementation](https://github.com/openai/CLIP/issues/83). When I changed the weight decay from 0.2 to 0.001, it works well on my task.

**Comment by iremonur** [comment]
Hi, I aim to finetune the clip but I reached %30 of accuracy at max with the learning rate of 5e-6 and weight decay of 0.2, respectively. How is your accuracy in your experiment? What is the size of your data which you use for fine-tuning? I would be glad if you help me with this.

**Comment by dongyunkim-arch** [comment]
@iremonur Hi! I am wondering how you get the accuracy metrics when training since I could only see the loss... Is there any other function or script I can refer to to plot acc metrics? Thanks!

**Comment by jweihe** [comment]
> I have finetuned CLIP on PASCAL VOC2012 dataset using this [implementation](https://github.com/openai/CLIP/issues/83). When I changed the weight decay from 0.2 to 0.001, it works well on my task.

Could you provide your fine-tune code?thank you~

**Comment by Sierkinhane** [comment]
My code was same with [the implementation](https://github.com/openai/CLIP/issues/83)

# Latest Update : 22 June 2021, 09:55 GMT+7

# TO ADD :
# Gradient Checkpointing
# Filter out bias from weight decay
# Decaying learning rate with cosine schedule
# Half-precision Adam statistics
# Half-precision stochastically rounded text encoder weights were used

#BATCH_SIZE must larger than 1
train_dataloader = DataLoader(...,batch_size = BATCH_SIZE) #Define your own dataloader

#https://github.com/openai/CLIP/issues/57
def convert_models_to_fp32(model): 
    for p in model.parameters(): 
        p.data = p.data.float() 
        p.grad.data = p.grad.data.float() 

device = "cuda:0" if torch.cuda.is_available() else "cpu" # If using GPU then use mixed precision training.
model, preprocess = clip.load("ViT-B/32",device=device,jit=False) #Must set jit=False for training
if device == "cpu":
  model.float()
else :
  clip.model.convert_weights(model) # Actually this line is unnecessary since clip by default already on float16

loss_img = nn.CrossEntropyLoss()
loss_txt = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=5e-5,betas=(0.9,0.98),eps=1e-6,weight_decay=0.2) #Params used from paper, the lr is smaller, more safe for fine tuning to new dataset

for epoch in range(EPOCH):
  for batch in train_dataloader :
      optimizer.zero_grad()

      list_image,list_txt = batch #list_images is list of image in numpy array(np.uint8), or list of PIL images
    
      images= torch.stack([preprocess(Image.fromarray(img)) for img in list_image],dim=0).to(device) # omit the Image.fromarray if the images already in PIL format, change this line to images=list_image if using preprocess inside the dataset class
      texts = clip.tokenize(list_txt).to(device)
    
      logits_per_image, logits_per_text = model(images, texts)

      ground_truth = torch.arange(BATCH_SIZE,dtype=torch.long,device=device)

      total_loss = (loss_img(logits_per_image,ground_truth) + loss_txt(logits_per_text,ground_truth))/2
      total_loss.backward()
      if device == "cpu":
         optimizer.step()
      else : 
        convert_models_to_fp32(model)
        optimizer.step()
        clip.model.convert_weights(model)

**Comment by jweihe** [comment]
> My code was same with [the implementation](https://github.com/openai/CLIP/issues/83)

ok， thank u

**Comment by ErinZhang1998** [comment]
@ABaldrati What do you mean that you freeze the normalization layers?

**Comment by sanjaygunda13** [comment]
> My code was same with [the implementation](https://github.com/openai/CLIP/issues/83)

I think this helps in training clip from scratch but not finetune. please let me know if i am wrong i am noob :p in all these stuff

**Comment by bwanglzu** [comment]
in case you're searching for some handy solutions, checkout this page https://finetuner.jina.ai/tasks/text-to-image/

no magic, help you get your job done.

**Comment by zzxslp** [comment]
> I have finetuned CLIP on PASCAL VOC2012 dataset using this [implementation](https://github.com/openai/CLIP/issues/83). When I changed the weight decay from 0.2 to 0.001, it works well on my task.

Hi, could you let me know the batch size you are using for fine-tuning CLIP? Thanks!

**Comment by zhentingqi** [comment]
Hi, any update in 2024? Is there any method to fine-tune the CLIP model using newest huggingface transformers library (4.35.*)? Thanks!

**Comment by tengshaofeng** [comment]
this is my training code, everything gose well. https://github.com/tengshaofeng/finetune-jina-clip-v2
